Modernize CMake build

wo80 opened the pull request:
This is a major rewrite of the CMake build system. CMake offers two ways of finding packages in a structured way - module mode and config mode (see cmake help [find_package](https://cmake.org/cmake/help/latest/command/find_package.html#search-modes)).

Both modes are now supported:
* Module mode using the FindXXX.cmake modules in the `cmake/modules` subdirectory. This mode is suited for finding dependencies installed by a package manager (either by looking for pkg-config files or directly looking for libs and headers).
* Config mode basically is CMake magic. This is especially useful when dependencies are installed using the shell scripts.

The shell scripts have been updated/streamlined. Both SuperLU and SuiteSparse now provide a CMake build system.

C++17 is required due to https://github.com/m-reuter/arpackpp/commit/db80c6d77de7eaef33bb3ee2e0464b1b7c2abf0a. `arerror.h` did not compile with Visual Studio due to the "trick to initialize static member code". This is fixed by using C++17 which lets you declare static data members as `inline`.

`README.md` and `INSTALL.md` have been updated.

I have tested this on Windows and Debian. Please take your time reviewing the changes - the pull request has gotten quite a bit larger than initially expected.

There are more changes in the https://github.com/wo80/arpackpp/tree/ctest branch. Examples are registered as tests, which can then be run using `ctest` (all tests pass on Debian). Let me know if those changes should be merged into this pull request or better go into a separate one.

wo80 committed:
Add CMake build directory to gitignore

The name 'build' seems to be the most common, so add it as default to gitignore.

wo80 committed:
Move cmake_minimum_required() to top of CMakeLists.txt and update minimum required version to 3.16

wo80 committed:
Update CMake project metadata and remove Fortran dependency

The Fortran language should not be a dependency of a C++ header only library. All dependencies which require Fortran will be included using CMake modules. ARPACK++ is not responsible for building these dependencies.

wo80 committed:
Fix arcomplex definition for MSVC compiler

wo80 committed:
Require C++17 and simplify ArpackError static initialization

wo80 committed:
Remove library suffixes setup and let CMake handle this

wo80 committed:
Simplify BLAS/LAPACK setup and remove explicit linking of gfortran libraries

wo80 committed:
Remove make_global_path and let CMake handle library paths

wo80 committed:
Cleanup top level CMakeLists.txt

* Move examples setup to subdirectory
* Add arpackpp dependency to examples
* Add CMake option to disable examples
* Minor cleanup of whitespace and capitalization

wo80 committed:
Add examples/CMakeLists.txt

wo80 committed:
Cleanup CMake setup of ARPACK and SuperLU

wo80 committed:
Use CMake find_package() to find SuperLU

wo80 committed:
Remove custom target 'Message'

wo80 committed:
Add CMake modules to find SuiteSparse packages

These are the official modules created when installing the current version of SuiteSparse (version >= 6.0)

wo80 committed:
Add ARPACK CMake module and cleanup top level CMakeLists.txt

wo80 committed:
Remove custom CMake modules for SuiteSparse.

As of SuiteSparse version 7.2 proper targets will be installed which can be picked up by CMake automatically.

wo80 committed:
Remove custom CMake module for ARPACK.

Recent versions of arpack-ng install proper CMake targets, so let's use them.

wo80 committed:
Update dependencies install scripts.

wo80 committed:
Update README.md

wo80 committed:
Update install scripts
* Make local install into ./external directory default (global install is available via -g or --global-install switch, which requires root access)
* Disable build of ARPACK tests
* Disable build of SuperLU examples and tests
* Explicitly specify release config build

wo80 committed:
Fix issues with CMake scripts
* Set CMAKE_PREFIX_PATH to enable config mode when installing locally into ./external directory
* Link arpackpp target to ARPACK to properly pass the link dependency

wo80 committed:
Add back Fortran to CMake in case ARPACK is linked statically.

wo80 committed:
Add back CMake modules.

This is necessary to find dependencies which were installed by a package manager and don't provide CMake config targets.

wo80 committed:
If CHOLMOD is enabled, add include directory.

wo80 committed:
Minor improvements to ARPACK and SuperLU CMake modules.

wo80 committed:
Update README.md and INSTALL.md

wo80 committed:
Update .travis.yml

wo80 commented:
I was expecting Travis to kick in... Maybe it's not configured to run on pull requests?

Anyway, what do you think about replacing Travis with a Github workflow?

m-reuter commented:
Wow thanks for all the work! This is great, love it. 
- I think the tests should be a separate PR.
- Not sure about travis. I totally agree moving to GitHub actions. 
- You probably did not test any of this on MAC ?

wo80 commented:
> Wow thanks for all the work! This is great, love it.

Thank you!

> You probably did not test any of this on MAC ?

I don't have a MAC. That's why having some CI enabled would come in handy :-)

> Not sure about travis. I totally agree moving to GitHub actions.

I'm not familiar with Travis, but <del>looking at the [docs](https://docs.travis-ci.com/user/pull-requests/)</del> - not the docs, but [stackoverflow](https://stackoverflow.com/questions/31882306/how-to-configure-travis-ci-to-build-pull-requests-merges-to-master-w-o-redunda) (though not sure, if that's still up to date), it seems there's a switch that must be enabled.

Actually, I could try set up a workflow on my ctest branch and see what's happening. Will play around with that tomorrow...

m-reuter commented:
Travis is enabled and has built PRs in the past. The reason now seems missing credits (although I still have some), their interface is not very clear about it. I will have to ask them.

wo80 commented:
Setting up the workflow was easy: https://github.com/wo80/arpackpp/blob/ctest/.github/workflows/build.yml

Now, while the workflow runs fine on ubuntu, macos fails. I don't think this has anything to do with the changes in this pull request, all the CMake steps (configure, build, test) are pocessed fine. The changes in the ctest branch don't affect the main arpackpp code either, so I suspect if the Travis CI had actually run any of the examples, the errors would have been there too.

If you look at https://github.com/wo80/arpackpp/actions/runs/6309846938/job/17130593414 there are three types of errors:
1. _Subprocess aborted_ in the  rcomp* rnsym* and rsym* examples, which throw an ArpackError exception.
2. _Subprocess aborted_ in lsvd and lsvd2, which is most likely caused by a memory management problem causing a segfault in arlnsmat.h. I already encountered this on Visual Studio (I'll open an issue for this).
3. _Failed_ tests due to lack of convergence

As mentioned, I don't have a Mac and since on Linux and Windows I cannot reproduce 1 and 3, I can't debug this. Maybe you have an idea?

m-reuter commented:
When trying this PR on my MacBook, installing all dependencies locally via the updated install scripts, I get this CMAKE error when ENABLE_SUPERLU is ON: 

`cmake -B build -D ENABLE_SUPERLU=ON  -D BLA_VENDOR=OpenBLAS`

> CMake Error in examples/CMakeLists.txt:
  Imported target "superlu::superlu" includes non-existent path
    "/usr/local/include"
>  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
>  * The path was deleted, renamed, or moved to another location.
>  * An install or uninstall procedure did not complete successfully.
>  * The installation package was faulty and references files it does not
  provide.

For Cholmod and Umfpack the examples build.

wo80 commented:
> When trying this PR on my MacBook, installing all dependencies locally via the updated install scripts, I get this CMAKE error when ENABLE_SUPERLU is ON

Can you post the complete output of the configure step? It should report 'Found XYZ' for each package and the path to the library should be displayed (see https://github.com/wo80/arpackpp/actions/runs/6309846938/job/17130593128#step:4:27). Is it the local one in the 'external' folder, or a path to a package installed globally?

If you look at the relevant code in CMakeLists.txt

https://github.com/wo80/arpackpp/blob/ef8231bd662258e69af4dfbb466cb7b6ffa0fae0/CMakeLists.txt#L25-L31

both paths for local install (config mode) and global packages (module mode) are specified and I rely on CMake choosing config mode automatically. You could try commenting out the `set(CMAKE_MODULE_PATH ...` line and see if that works.

Which version of CMake is installed?

m-reuter commented:
cmake version 3.27.6

It is in the 'external' folder.

Commenting out that line works! It now also builds all superLU examples, I ran a few and they look good. Now we just need to fix that CMAKE in a way that it does not break it for global installs. And then we can look at the GitHub workflow next. I think there you may not need to specifically specify the gcc . I did not do that either. 


Here is the full output from before:

cmake -B build -D ENABLE_SUPERLU=ON  -D BLA_VENDOR=OpenBLAS
-- The C compiler identification is AppleClang 14.0.3.14030022
-- The CXX compiler identification is AppleClang 14.0.3.14030022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libopenblas.a  
-- Looking for cheev_
-- Looking for cheev_ - found
-- Found LAPACK: /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libopenblas.a;/Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libopenblas.a  
-- Found arpackng: /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libarpack.dylib (found version "3.9.0") 
-- Found superlu: /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libsuperlu.a (found version "6.0.1") 
-- Configuring done (0.8s)
CMake Error in examples/CMakeLists.txt:
  Imported target "superlu::superlu" includes non-existent path
    "/usr/local/include"
  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
  * The path was deleted, renamed, or moved to another location.
  * An install or uninstall procedure did not complete successfully.
  * The installation package was faulty and references files it does not
  provide.


This error keeps repeating for many times and then it say:

-- Generating done (0.3s)
CMake Generate step failed.  Build files cannot be regenerated correctly.

wo80 commented:
Alright, two things to try separately (don't forget to undo the commenting of the line):

1. In `install-superlu.sh` change

   `cmake -B build -D enable_examples=OFF -D enable_tests=OFF` to
   `cmake -B build -D enable_examples=OFF -D enable_tests=OFF -D CMAKE_INSTALL_PREFIX="$(pwd)/external"`
2. In `cmake/modules/Findsuperlu.cmake` remove the `PATHS` hints from `find_path(...)`. I just tested this on Debian and CMake is clever enough to find the package anyway.

Since commenting out the `CMAKE_MODULE_PATH` worked, 1) shouldn't make a difference. But I think that change should be made to all the install scripts, just to make sure all the targets pick up the correct install prefix during the configure step.

wo80 commented:
> And then we can look at the GitHub workflow next. I think there you may not need to specifically specify the gcc .

It seems, though I tried to specify GCC explicitly, CMake is still choosing clang for the macos job:
```
-- The C compiler identification is AppleClang 14.0.0.14000029
-- The CXX compiler identification is AppleClang 14.0.0.14000029
[...]
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Check for working CXX compiler: /usr/bin/g++ - skipped
```
Not sure what to make out of that...

m-reuter commented:
yes, gcc is just sym linked to clang stuff. But clang works nicely too (most of the time).

m-reuter commented:
I did those edits, build superLU again,  but get the same errors. I remove the PATH in two places in Findsuperlu:

  find_path(SUPERLU_INCLUDE_DIRS
    NAMES slu_util.h
    PATH_SUFFIXES
      superlu)

  find_library(SUPERLU_LIBRARIES
    NAMES superlu)

wo80 commented:
Is the error message still about `/usr/local/include`? That path should not be in any of the CMake files (edit: any files relevant to SuperLU), so the problem must be somewhere else.

Another try (just a guess resulting from a google search):

Undo the above changes and in [examples/CMakeLists.txt](https://github.com/wo80/arpackpp/blob/ef8231bd662258e69af4dfbb466cb7b6ffa0fae0/examples/CMakeLists.txt#L27-L34) change the linking in all calls to `target_link_libraries(...)` from `PUBLIC` to `PRIVATE`. I think this is the right thing to do for executables anyway.

m-reuter commented:
Actually it was a different error:

CMake Error in examples/CMakeLists.txt:
  Imported target "superlu::superlu" includes non-existent path

    "/Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/superlu/external/include"

m-reuter commented:
OK, I solved that by doing:
`install_dir="$(pwd)/external"`
in the top of install_superlu.sh 
and then
`cmake -B build -D enable_examples=OFF -D enable_tests=OFF -D CMAKE_INSTALL_PREFIX="$install_dir"`

Then it configures without error after re-installing superLU. But building the examples does not work this way due to duplicate symbol:

> duplicate symbol '_z_abs' in:
    /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libopenblas.a(z_abs.c.o)
    /Users/martin/SynologyDrive/projects/git/arpack-test/arpackpp/external/lib/libsuperlu.a(dcomplex.c.o)

m-reuter commented:
Also switching from PUBLIC to PRIVATE in does not change that.

Probably superLU needs to be build knowing about the BLAS lib?

wo80 commented:
Just for clarification:
* `CMAKE_INSTALL_PREFIX="$(pwd)/external"` doesn't work
* `CMAKE_INSTALL_PREFIX="$install_dir"` with `install_dir="$(pwd)/external"` works?

> Also switching from PUBLIC to PRIVATE in does not change that.

Thanks for trying, I think I'll add that change anyway.

> Probably superLU needs to be build knowing about the BLAS lib?

I don't think so, the duplicate symbols are in SuperLU's `dcomplex.c`, which is included by default. I see two possible solutions:
1) compile and then link either OpenBLAS or SuperLU as a shared library
2) ask SuperLU to rename the symbols
3) EDIT: apply a patch here to rename the symbols in the SuperLU source code.

m-reuter commented:
> Just for clarification:
> CMAKE_INSTALL_PREFIX="$(pwd)/external" doesn't work
> CMAKE_INSTALL_PREFIX="$install_dir" with install_dir="$(pwd)/external" works?

Yes, the reason is that we change into external and into superlu so the first will not work as pwd is in the wrong place. Therefore storing the install_dir at the beginning of the script and then using it later works. 

About the duplicate symbol, I did not get those errors earlier when I commented out the ` set(CMAKE_MODULE_PATH ... `line

wo80 commented:
I have to admit the one thing I didn't test is building OpenBLAS locally. I'm currently stuck at
```
cmake -B build -D ENABLE_SUPERLU=ON -D BLA_VENDOR=OpenBLAS
[...]
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /home/christian/arpackpp-cmake/external/lib/libopenblas.a  
-- Looking for cheev_
-- Looking for cheev_ - not found
CMake Error at /usr/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find LAPACK (missing: LAPACK_LIBRARIES)
```
Very strange that BLAS is found but LAPACK isn't. Overall, I think that the install scripts might not pick up the locally installed BLAS provider in case another one is installed on the system. This needs more testing...

---

Here's another idea to improve matters with `set(CMAKE_MODULE_PATH ...)`:

I think CMake is preferring module mode and thus - in case `CMAKE_MODULE_PATH` points to the right location - it always chooses the `FindXYZ` modules in `cmake/modules` instead of the locally installed config targets. To fix this, two calls to `find_package` might be needed (for all packages). Something like
```
find_package(superlu NO_MODULE)
  
if (NOT TARGET superlu::superlu)
  find_package(superlu REQUIRED)
endif()
```
The `NO_MODULE` option instructs CMake to run in config mode exclusively. In case no config is found, we retry with module mode.

wo80 committed:
Improve CMake find_package usage.

Use a custom function which first tries pure config mode and in case the package isn't found also allows module mode.

wo80 committed:
Specify link libraries for examples as PRIVATE.

wo80 committed:
Set CMAKE_INSTALL_PREFIX in install scripts to make sure that the correct prefix is set during configuration.

wo80 committed:
Remove PATHS hint from superlu and arpackng CMake modules.

wo80 commented:
I rebased the ctest branch to include the latest changes. The ubuntu job still works, the macos job still fails.

I have no idea why the Travis CI build fails:
* The linux job uses xenial, which is rather outdated, so I don't know what the arpack package looks like, but the error
  `Could NOT find arpackng (missing: ARPACK_INCLUDE_DIRS) (found version "")`
  indicates that `arpack.h` isn't found. It should be possible to fix that by using the install script, but in case you plan to replace Travis by a Github workflow, I wonder if it makes sense trying to fix the problems with Travis?
* The macos job fails with `No output has been received in the last 10m0s`, so that's definitely a problem on the Travis side.

I haven't solved the problem using OpenBLAS as a static library yet. Building it as a shared library works, though, and - as expected - linking against the shared library, there are no conflicts with duplicate symbols in SuperLU.

But I can confirm now that building OpenBLAS locally and then using the install scripts, the local OpenBLAS lib is not used. For SuperLU that could be fixed by adding `-D TPL_BLAS_LIBRARIES=$(pwd)/external/lib/libopenblas.a`, but haven't found a way to do this for arpack-ng and SuiteSparse.

So, if arpack-ng and SuiteSparse don't support linking to a local BLAS lib using CMake (I don't know, if that's the case), the question is, whether arpackpp should still try to support that?

EDIT:
Specifying
```
-D BLAS_LIBRARIES=$(pwd)/external/lib/libopenblas.a
-D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a
```
seems to work for arpack-ng and SuiteSparse. So the shell scripts can check whether that path exists and then pass it to CMake. I'll fix it tomorrow...

m-reuter commented:
The reason we have the install scripts and support for static libraries:
- I did not want to change the global system by installing distribution packages. 
- distribution packages have been extremely outdated for a long time, so in order to get a decent version it had to be compiled or installed directly. Not sure how much that has improved. 
- Some distribution packages did not provide static libraries, if I remember correctly.
- I needed static for easier distribution of a pre-compiled binary that included everything, so that it can be used without root access to install dependencies (and without having to distribute and setup dynamic libraries).

So we need to support static builds, and I'd like to continue support for local builds of dependencies, for users who cannot or do not want to change their system globally via package installer.

m-reuter commented:
I tested this new version on my Mac and installing the dependencies and building everything works now. I think, however, it is using the dynamic libraries. Is there a cmake flag to tell it to use static?

wo80 commented:
> Is there a cmake flag to tell it to use static?

Yes, CMake has the `BUILD_SHARED_LIBS` flag, which is `OFF` by default.
* arpack-ng sets this `ON`, so it has to be passed on the command line `-D BUILD_SHARED_LIBS=OFF`.
* SuperLU uses the default.
* SuiteSparse unfortunately has it's own way own handling this. The `BUILD_SHARED_LIBS` flag is ignored completely and instead both shared and static libraries are build by default, for example `amd.lib` and `amd_static.lib`. At the moment, only the shared libs are considered when configuring arpackpp. We cannot just switch to the static version, because that would break using packages installed on the system. So I'll have to think about, how that can be accomplished (though, to be fair, the old CMake script wouldn't have worked too (edit: with the recent changes in the SuiteSparse build system), so you decide, if SuiteSparse should be a priority now, or if that can be delayed).

wo80 commented:
Is the problem with duplicate symbols still present?

m-reuter commented:
> Is the problem with duplicate symbols still present?

No, that disappeared. Not sure why. 

And for the SuperLU examples, it only included the arpack as dynamic library. For cholmod and umfpack those libraries respectively (and arpack) are also linked dynamically by default. I would need to check what the behaviour was before, but I thought we always build completely static (if the .a files are there in the external directory). Either way it should be consistent for all libraries (e.g. blas and superLU also) and it should agree with the setting in BUILD_SHARED_LIBS. So if that is off by default, then it should build completely static. If a static dependency is missing it should fail.

wo80 commented:
I'll push the changes to the install scripts later this day. I think adding BUILD_SHARED_LIBS=OFF to all of them is the best solution (even if not necessary).

Fixing the usage of SuiteSparse might take a bit longer.

wo80 commented:
Small update:

It seems SuiteSparse does respect the `BUILD_SHARED_LIBS` flag. Though the build still produces shared object libraries, the .a library seems to be a proper static import library:
```
dd ./usymgreg
	linux-vdso.so.1 (0x00007fff691dd000)
	libflexiblas.so.3 => /usr/local/lib/libflexiblas.so.3 (0x00007ffa94800000)
	libarpack.so.2 => /home/christian/arpackpp-cmake/external/lib/libarpack.so.2 (0x00007ffa94c91000)
	libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007ffa94c34000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffa94721000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ffa94400000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ffa94c12000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffa9421f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ffa94db7000)
	libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007ffa93e00000)
	libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007ffa94bcb000)
```

All references to SuiteSparse libraries in the CMake build script have to be changed from `SuiteSparse::LIBNAME` to `SuiteSparse::LIBNAME_static`. That should be easy using CMake generator expressions. I'm thinking of the following solution:

* Introduce another build option `PREFER_SUITESPARSE_STATIC` (default `OFF`)
* Use generator expressions wherever necessary, like
  `$<IF:PREFER_SUITESPARSE_STATIC,SuiteSparse::AMD_static,SuiteSparse::AMD>`

wo80 committed:
Update install scripts to enable static build.
* always specify BUILD_SHARED_LIBS=OFF
* explicitly link to local static openblas.a if present

wo80 committed:
Enable linking to static SuiteSparse CMake targets.

wo80 commented:
I think we're getting closer:
```
ldd ./usymgreg
	linux-vdso.so.1 (0x00007ffe19c8a000)
	libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007fc444ccb000)
	libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007fc444a00000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc444600000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc444921000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc444901000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc44441f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc444e79000)
	libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007fc4448b8000)
```
The CMake command was
```
cmake -B build 
      -D ENABLE_SUPERLU=ON 
      -D ENABLE_CHOLMOD=ON 
      -D ENABLE_UMFPACK=ON 
      -D PREFER_SUITESPARSE_STATIC=ON 
      -D BLAS_LIBRARIES=$(pwd)/external/lib/libopenblas.a 
      -D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a
```

wo80 committed:
Update .travis.yml to use latest os versions.

m-reuter commented:
I think it is best, if the install shell scripts build both shared and static libs for the dependencies in external. But our cmake should have a single flag to specify if users want to build agains shared libs of the dependencies or against static. It will then search for the selected ones in "external" and then globally.

wo80 commented:
> I think it is best, if the install shell scripts build both shared and static libs for the dependencies in external.

The shell scripts now all have `build_type="-D BUILD_SHARED_LIBS=OFF"` on top. We could either leave it that way and let the user edit the scripts in case they want to have shared libs, or - for convenience - add an option so that, for example
`./install_xyz.sh --shared-libs`
switches that to `BUILD_SHARED_LIBS=ON`.

> But our cmake should have a single flag to specify if users want to build agains shared libs of the dependencies or against static. It will then search for the selected ones in "external" and then globally.

I'm not sure if it's worth the trouble implementing that. The CMake script should work with what it finds in the external folder. The user decides what he wants by choosing the option from the install script - if used at all.
To be honest, I think that using the packages provided by a distribution can be regarded the _most common_ way how to build a library like this. After all, arpackpp is a header-only library. At this point we have a build system that works both with system packages and local install scripts (which I see more as a bonus).

FYI, I might not be as responsive the next couple of days.

m-reuter commented:
Thanks, again for all your great work. I think we can merge this and work on details later if really needed. I agree that most users will use the system libraries, but they are very old. SuiteSparse is distributed as 5.10 in Ubuntu 22.04, while the current version is 7.2. SuperLU is distributed as 5.3, while latest is 6.0.1. So we should probably keep the local builds working (and maybe include them into our CI tests also). 

Not sure why the Travis Linux is timing out. Will try a local build on Ubuntu 22.04 and on Centos tomorrow.

wo80 commented:
> I think we can merge this and work on details later if really needed.

I think adding a --shared-libs option to the install scripts makes sense. I'm also wondering, whether it's wise to clone/pull from master. This might break things sooner or later. In case we decide to clone specific tags, we could also switch to downloading them as archives using wget or curl. That should be faster than using git. The versions could then be updated from time to time, once they are tested to work.

> So we should probably keep the local builds working (and maybe include them into our CI tests also).

I agree. But that can be addressed in the ctest pull request, once you decide whether you want to keep Travis or not (I think it's pretty annoying).

As already mentioned, I'm not sure how much time I have to work on this in the coming days. So maybe the next commit won't be before next week.

m-reuter commented:
on my ubuntu 22.04 I installed dependencies via the install scripts, and liblapack-dev via apt and it does not find lapack during the cmake build. Also in your INSTALL.md I do not see how to install Lapack.

wo80 commented:
> it does not find lapack during the cmake build

Please clarify: do you mean the CMake configure step completes and the actual building of the examples fails, or does the configure step fail with an error message that LAPACK isn't found? If the latter, what is the CMake output regarding BLAS and LAPACK.

At the moment, I completely rely on the CMake built-in [FindBLAS](https://cmake.org/cmake/help/latest/module/FindBLAS.html) / [FindLAPACK](https://cmake.org/cmake/help/latest/module/FindLAPACK.html) modules and it's a bit surprising that it doesn't find the package. I have never cared about the generic LAPACK, always used OpenBLAS, and recently trying FlexiBLAS. You might try setting `-D BLA_VENDOR=Generic`.

So, if I understand correctly, you have installed liblapack-dev only, no other BLAS/LAPACK provider? That should be supported out of the box, but I never tried. I'll do so when I find some time.

EDIT:

> Also in your INSTALL.md I do not see how to install Lapack.

The INSTALL.md will also need some updates before merging this, but I haven't mentioned LAPACK there because it shouldn't be necessary to install when using OpenBLAS.

m-reuter commented:
Yes, I install liblapack-dev and nothing else. Openblas was compiled using the install script. The configure failed complaining that Lapack was not found. I can post the output tomorrow. Will probably start trying these things in a Docker to make sure nothing else is messing things up.

wo80 commented:
So you are not trying to build against liblapack-dev, but the local openblas? Then maybe you are running into the same error I mentioned in https://github.com/m-reuter/arpackpp/pull/27#issuecomment-1735671103

I haven't found out what's the cause. The only way to fix it, was to explicitly state the libraries to build against:
```
-D BLAS_LIBRARIES=$(pwd)/external/lib/libopenblas.a 
-D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a
```
Since BLAS seems to be picked up correctly, maybe specifying only LAPACK is enough. The question remains, why BLAS is found, but LAPACK isn't ... makes little sense.

m-reuter commented:
Yes, I tried to build with local openblas (external dir) but I got the error that LAPACK was not found. I then installed liblapack-dev via apt and it still did not find it. I had specified `BLA_VENDOR=OpenBLAS` as per your INSTALL documentation.

m-reuter commented:
I now added 
`-D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a`
and it configures and compiles everything. However for some examples, it links with dynamic libraries from SuiteSparse and then does not find them on execution:

```
ldd ./usymgshf 
	linux-vdso.so.1 (0x00007ffcabf91000)
	libumfpack.so.6 => /groups/ag-reuter/users/reuter/arpackpp/external/lib/libumfpack.so.6 (0x00007f8d9ae30000)
	libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f8d9aa00000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8d9a600000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8d9ad36000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8d9ad16000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8d9a200000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8d9b089000)
	libcholmod.so.4 => not found
	libamd.so.3 => not found
	libsuitesparseconfig.so.7 => not found
	libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f8d9a9b8000)
```

m-reuter commented:
> I think adding a --shared-libs option to the install scripts makes sense. I'm also wondering, whether it's wise to clone/pull from master. This might break things sooner or later. In case we decide to clone specific tags, we could also switch to downloading them as archives using wget or curl. That should be faster than using git. The versions could then be updated from time to time, once they are tested to work.

I agree with both. Also the old install scripts mainly downloaded release versions (I think except for aprack-ng, not sure). It would be good to use curl and install releases instead of git.

wo80 commented:
I haven't looked in detail at the `FindBLAS` / `FindLAPACK` modules that CMake provides, but it might be the case that they primarily look for official package installs, so specifying `BLA_VENDOR=OpenBLAS` might not be right thing to do to pick up the library in the `external` folder. What I don't understand, is why BLAS is found, but LAPACK isn't. It might be worth letting the CMake developers know.

Anyway, here's an idea, how to circumvent the problem:
* Introduce another CMake option, like `ENABLE_EXTRNAL_DIR` or `ENABLE_LOCAL_LIBS`, default = `OFF`.
* If that option is `ON`, before calling `find_package(BLAS)` and `find_package(LAPACK)`, do
   ```
   if(ENABLE_LOCAL_LIBS AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/external/lib/libopenblas.a")
      set(BLAS_LIBRARIES "${CMAKE_CURRENT_SOURCE_DIR}/external/lib/libopenblas.a")
      set(LAPACK_LIBRARIES "${CMAKE_CURRENT_SOURCE_DIR}/external/lib/libopenblas.a")
   endif()
   ```
If setting `-D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a` on the command line works, that should work too.

The other option is to leave everything as is and update the INSTALL.md: instead of using `-D BLA_VENDOR=OpenBLAS` use  `-D LAPACK_LIBRARIES=$(pwd)/external/lib/libopenblas.a` (of course needs testing, that this works reliably).

wo80 commented:
> However for some examples, it links with dynamic libraries from SuiteSparse and then does not find them on execution:

Have you passed `-D PREFER_SUITESPARSE_STATIC=ON`? If not, CMake links against the shared libraries, and since `external/lib` is not in your `PATH`, the programs won't run.

That also needs to be documented. And I'm thinking about renaming that option to `ENABLE_SUITESPARSE_STATIC`. Looks nicer.

wo80 commented:
I will be away from my PC until Sunday. If you find some time, use the Github review feature and add reviews to all the code places that you think need updates :-)

m-reuter commented:
Just found out that the `-D BLA_VENDOR=OpenBLAS` was actually causing the problem. Without it it finds everything in external. So that should be solved by explaining better when the BLA_VENDOR OpenBLAS should really be used. I thought always when we want the local OpenBlas, but probably not. Still the suitesparse dynamic issue however remained.

That disappeared by adding `-D PREFER_SUITESPARSE_STATIC=ON`  as you said, but I don't like to have to do that. And then only for SuiteSparse while the others work, that is not consistent. 

Since often both static and dynamic libraries might be installed (e.g. on the system) I think the behaviour should be:
- install scripts should always build/install both (if available) into external directory.
- our cmake should default to using the static first (or the dynamic, I don't care about the default). It should be possible to ask for the other specifically via a flag and get a failure if those libraries are not available. 
- our cmake should always prefer the external libraries if they exist over system libraries by default.

wo80 commented:
> That disappeared by adding -D PREFER_SUITESPARSE_STATIC=ON as you said, but I don't like to have to do that. And then only for SuiteSparse while the others work, that is not consistent.

You are absolutely right and I don't like that, too. But that's what SuiteSparse offers to us, as mentioned in https://github.com/m-reuter/arpackpp/pull/27#issuecomment-1736960041

Their CMake build system is in constant shift, many changes between 6.0 and current version, so maybe in the future, they will offer a more consistent way to switch between static and shared, but right now, the only way to do so is explicitly choosing between `CHOLMOD` and `CHOLMOD_static` targets (and the like).

m-reuter commented:
I thought suitesparse builds both by default (.so and .a). Then our cmake scripts need to know which ones to include, and that can be done via a static or shared request variable in our script that then makes sure that the right library is searched for?

wo80 committed:
Update install scripts
* Add option to build shared libs
* Explicitly specify build type (default is Release)
* Instead of using git master branch, download specific tag using wget

wo80 committed:
In case SuiteSparse static libs are available, set CMake option ENABLE_SUITESPARSE_STATIC to true.

wo80 commented:
The install scripts should be pretty much complete by now. Be aware that arpack-ng 3.9.0 and SuperLU 6.0.1 have some known bugs (both already fixed in master), so the scripts should be updated, when patch versions are released (the bugs are not relevant to arpackpp, as far as I can see). I decided to use the beta release of SuiteSparse, because - as mentioned above - their CMake build system has changed a lot lately and this version seems to work fine so far. Since we are not in a hurry, waiting for new releases of these dependencies before merging is an option...

The tests below show that the normal use cases work fine (on Debian): build with system packages only and build with local static packages. I don't see a way how to completely get rid of the option `ENABLE_SUITESPARSE_STATIC`, but the way it is implemented now, the use case "build with local static libraries" is supported without having to specify any extra options.

In general, I think building against shared libraries which only exist in the 'external' folder doesn't make much sense. So that use case (using shared libraries built with the install scripts) is only supported if those libraries are installed globally.

I don't think it is important to implement the use case "build against system packages, though local packages in 'external' folder are present". It's as easy as temporarily renaming or deleting the 'external' folder. If you think it is important to have that functionality in the CMake build system, it should be possible by just not adding the 'external' folder to the [CMake prefix paths](https://github.com/wo80/arpackpp/blob/9a24ea3a614dfd992f1ca0745e593c31f2aad1bd/CMakeLists.txt#L33-L35). That would need another config option.

---

### Debian tests

Everything worked as expected on a fresh Debian virtual machine. Note that I haven't addressed the `LAPACK not found` error. I don't know what's the cause of this error, but (surprisingly), it didn't show up in the following tests.

I have stripped most of the output, only the parts relevant to the found libraries are shown:

```
# Using generic BLAS/LAPACK system packages

sudo apt install -y gfortran libblas-dev liblapack-dev libarpack2-dev
cmake -B build
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/liblapack.so;/usr/lib/x86_64-linux-gnu/libblas.so
-- Found arpackng: /usr/lib/x86_64-linux-gnu/libarpack.so

# Using OpenBLAS system package

sudo apt install -y libopenblas-dev
cmake -B build
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl
-- Found arpackng: /usr/lib/x86_64-linux-gnu/libarpack.so

# Using generic BLAS/LAPACK, with OpenBLAS installed

cmake -B build -DBLA_VENDOR=Generic
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/liblapack.so;/usr/lib/x86_64-linux-gnu/libblas.so
-- Found arpackng: /usr/lib/x86_64-linux-gnu/libarpack.so

# Using SuperLU and SuiteSparse system packages

sudo apt install -y libsuperlu-dev libsuitesparse-dev
cmake -B build -DENABLE_SUPERLU=ON -DENABLE_CHOLMOD=ON -DENABLE_UMFPACK=ON
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl
-- Found arpackng: /usr/lib/x86_64-linux-gnu/libarpack.so
-- Found superlu [config]: /usr/lib/x86_64-linux-gnu/libsuperlu.so.5.3.0
-- Found SuiteSparse_config: /usr/lib/x86_64-linux-gnu/libsuitesparseconfig.so.5.12.0 (found version "5.12.0") 
-- Found AMD: /usr/lib/x86_64-linux-gnu/libamd.so.2.4.6 (found version "2.4.6") 
-- Found CAMD: /usr/lib/x86_64-linux-gnu/libcamd.so.2.4.6 (found version "2.4.6") 
-- Found CCOLAMD: /usr/lib/x86_64-linux-gnu/libccolamd.so.2.9.6 (found version "2.9.6") 
-- Found COLAMD: /usr/lib/x86_64-linux-gnu/libcolamd.so.2.9.6 (found version "2.9.6") 
-- Found CHOLMOD: /usr/lib/x86_64-linux-gnu/libcholmod.so.3.0.14 (found version "3.0.14") 
-- Found UMFPACK: /usr/lib/x86_64-linux-gnu/libumfpack.so.5.7.9 (found version "5.7.9")

# Using arpack-ng locally

./install-arpack-ng.sh
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl
cmake -B build
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl
-- Found arpackng [config]: /home/christian/arpackpp-cmake/external/lib/libarpack.a

# Using arpack-ng + SuperLU locally

./install-superlu.sh
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
cmake -B build
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl
-- Found arpackng [config]: /home/christian/arpackpp-cmake/external/lib/libarpack.a
-- Found superlu [config]: /home/christian/arpackpp-cmake/external/lib/libsuperlu.a

# Using arpack-ng + SuperLU + SuiteSparse locally

./install-suitesparse.sh
cmake -B build
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl  
-- Found arpackng [config]: /home/christian/arpackpp-cmake/external/lib/libarpack.a
-- Found superlu [config]: /home/christian/arpackpp-cmake/external/lib/libsuperlu.a
-- Linking SuiteSparse static
-- Found SuiteSparse_config: /home/christian/arpackpp-cmake/external/lib/libsuitesparseconfig.so.7.2.1 (found version "7.2.1")
-- Found AMD [config]: /home/christian/arpackpp-cmake/external/lib/libamd.a
-- Found CAMD [config]: /home/christian/arpackpp-cmake/external/lib/libcamd.a
-- Found CCOLAMD [config]: /home/christian/arpackpp-cmake/external/lib/libccolamd.a
-- Found COLAMD [config]: /home/christian/arpackpp-cmake/external/lib/libcolamd.a
-- Found CHOLMOD [config]: /home/christian/arpackpp-cmake/external/lib/libcholmod.a
-- Found UMFPACK [config]: /home/christian/arpackpp-cmake/external/lib/libumfpack.a

# Build OpenBLAS locally (rebuild arpack-ng, SuperLU, SuiteSparse to use local OpenBLAS)

./install-openblas.sh
./install-arpack-ng.sh
-- Found BLAS: /home/christian/arpackpp-cmake/external/lib/libopenblas.a
-- Found LAPACK: /home/christian/arpackpp-cmake/external/lib/libopenblas.a
./install-superlu.sh
-- Found BLAS: /home/christian/arpackpp-cmake/external/lib/libopenblas.a
./install-suitesparse.sh
-- Found BLAS: /home/christian/arpackpp-cmake/external/lib/libopenblas.a
-- Found LAPACK: /home/christian/arpackpp-cmake/external/lib/libopenblas.a

cmake -B build
-- Found BLAS: /home/christian/arpackpp-cmake/external/lib/libopenblas.a
-- Found LAPACK: /home/christian/arpackpp-cmake/external/lib/libopenblas.a;-lm;-ldl
-- Found arpackng [config]: /home/christian/arpackpp-cmake/external/lib/libarpack.a
-- Found superlu [config]: /home/christian/arpackpp-cmake/external/lib/libsuperlu.a
-- Linking SuiteSparse static
-- Found SuiteSparse_config: /home/christian/arpackpp-cmake/external/lib/libsuitesparseconfig.so.7.2.1 (found version "7.2.1")
-- Found AMD [config]: /home/christian/arpackpp-cmake/external/lib/libamd.a
-- Found CAMD [config]: /home/christian/arpackpp-cmake/external/lib/libcamd.a
-- Found CCOLAMD [config]: /home/christian/arpackpp-cmake/external/lib/libccolamd.a
-- Found COLAMD [config]: /home/christian/arpackpp-cmake/external/lib/libcolamd.a
-- Found CHOLMOD [config]: /home/christian/arpackpp-cmake/external/lib/libcholmod.a
-- Found UMFPACK [config]: /home/christian/arpackpp-cmake/external/lib/libumfpack.a
```

wo80 committed:
Update .travis.yml

m-reuter commented:
thanks for all the work. For the install scripts we should use curl as it is available by default on Mac (while wget isn't).

m-reuter commented:
I also agree with your comments above about not linking against shared in external folder and not linking against global if the external folder exists. 
Other than the curl everything looks good to me. I may not have the time to test it in the coming weeks due to travel. I think we should not wait for new releases of dependencies and merge this soon (basically after the switch to curl). We can bump the versions of dependencies again before we do our next release. In the meantime we can switch to GitHub workflow and automated testing.

wo80 committed:
Update install scripts (support both curl and wget).

wo80 committed:
Update CHANGES.md

wo80 committed:
Update INSTALL.md

wo80 commented:
Please review the latest changes and then feel free to merge. I will open a pull request for the ctest stuff after this is merged, but finishing that can wait in case you are busy.

m-reuter commented:
Thanks for the great work, looks good.
